### ISSUE TITLE
Bug: field names are not returned correctly via gRPC

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#491]
__Steps to reproduce:__
1. Configure telegraf to send data to local IOx server
2. Start up local cluster and use UI to explore the data
3. Pick the CPU measurement

__Expected behaviour:__
<img width="1211" alt="Screen Shot 2020-11-25 at 3 06 03 PM" src="https://user-images.githubusercontent.com/490673/100276781-15772e80-2f30-11eb-9344-c63671b806ce.png">

The fields should be:

```
usage_system
usage_idle
usage_irq
usage_softirq
usage_steal
usage_guest
usage_guest_nice
usage_user
usage_nice
usage_iowait
```

__Actual behaviour:__
No field names are displayed

__Logs:__

The following call is logged:
```
[2020-11-25T20:09:07Z INFO  influxdb_iox::server::rpc::storage] tag_keys for database 26f7e5a4b7be365b_917b97a92e883afc, range: Some(TimestampRange { start: 1606334647520195000, end: 1606334947520195000 }) 
```

__Example Line Protocol Data__
```
mem,host=MacBook-Pro.hsd1.ma.comcast.net used_percent=47.16717004776001,inactive=14540857344i,wired=5720952832i,used=32413032448i,available=36306444288i,available_percent=52.83282995223999,active=22328414208i,free=21765586944i,total=68719476736i 1606335090000000000
cpu,cpu=cpu0,host=MacBook-Pro.hsd1.ma.comcast.net usage_nice=0,usage_softirq=0,usage_guest_nice=0,usage_system=10,usage_idle=79.40000000002328,usage_iowait=0,usage_irq=0,usage_steal=0,usage_guest=0,usage_user=10.600000000013097 1606335090000000000
cpu,cpu=cpu1,host=MacBook-Pro.hsd1.ma.comcast.net usage_softirq=0,usage_steal=0,usage_guest=0,usage_system=0.299999999998854,usage_idle=99.59999999991967,usage_nice=0,usage_irq=0,usage_user=0.09999999999734427,usage_iowait=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu2,host=MacBook-Pro.hsd1.ma.comcast.net usage_steal=0,usage_user=7.099999999970605,usage_system=8.299999999975116,usage_softirq=0,usage_irq=0,usage_guest=0,usage_guest_nice=0,usage_idle=84.59999999981781,usage_nice=0,usage_iowait=0 1606335090000000000
cpu,cpu=cpu3,host=MacBook-Pro.hsd1.ma.comcast.net usage_system=0.20000000000095497,usage_idle=99.70000000001164,usage_nice=0,usage_guest=0,usage_user=0.09999999999763531,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_steal=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu4,host=MacBook-Pro.hsd1.ma.comcast.net usage_idle=88.888888888986,usage_irq=0,usage_steal=0,usage_guest=0,usage_softirq=0,usage_guest_nice=0,usage_user=5.505505505519394,usage_system=5.605605605622064,usage_nice=0,usage_iowait=0 1606335090000000000
cpu,cpu=cpu5,host=MacBook-Pro.hsd1.ma.comcast.net usage_softirq=0,usage_guest=0,usage_user=0.0999000998979351,usage_idle=99.70029970031074,usage_iowait=0,usage_irq=0,usage_steal=0,usage_guest_nice=0,usage_system=0.1998001998004131,usage_nice=0 1606335090000000000
disk,device=disk1s1,fstype=apfs,host=MacBook-Pro.hsd1.ma.comcast.net,mode=ro,path=/ used=11240935424i,used_percent=9.654295691227624,inodes_total=9767978160i,inodes_free=9767489767i,inodes_used=488393i,total=1000240963584i,free=105193611264i 1606335090000000000
cpu,cpu=cpu6,host=MacBook-Pro.hsd1.ma.comcast.net usage_system=3.5964035964060654,usage_idle=92.80719280718787,usage_nice=0,usage_steal=0,usage_guest=0,usage_user=3.5964035964060654,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu7,host=MacBook-Pro.hsd1.ma.comcast.net usage_nice=0,usage_irq=0,usage_steal=0,usage_guest_nice=0,usage_user=0.09999999999734427,usage_idle=99.79999999981432,usage_softirq=0,usage_guest=0,usage_system=0.099999999999618,usage_iowait=0 1606335090000000000
cpu,cpu=cpu8,host=MacBook-Pro.hsd1.ma.comcast.net usage_user=2.897102897108924,usage_nice=0,usage_iowait=0,usage_softirq=0,usage_steal=0,usage_system=2.8971028970998383,usage_idle=94.20579420592753,usage_irq=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
disk,device=disk1s2,fstype=apfs,host=MacBook-Pro.hsd1.ma.comcast.net,mode=rw,path=/System/Volumes/Data inodes_used=3848878i,total=1000240963584i,free=105193611264i,used=877618663424i,used_percent=89.29667302971013,inodes_total=9767978160i,inodes_free=9764129282i 1606335090000000000
cpu,cpu=cpu9,host=MacBook-Pro.hsd1.ma.comcast.net usage_system=0.10000000000047748,usage_idle=99.89999999990687,usage_nice=0,usage_softirq=0,usage_guest_nice=0,usage_user=0,usage_iowait=0,usage_irq=0,usage_steal=0,usage_guest=0 1606335090000000000
disk,device=disk1s5,fstype=apfs,host=MacBook-Pro.hsd1.ma.comcast.net,mode=rw,path=/private/var/vm used=5368807424i,used_percent=4.855906272411087,inodes_total=9767978160i,inodes_free=9767978155i,inodes_used=5i,total=1000240963584i,free=105193611264i 1606335090000000000
cpu,cpu=cpu10,host=MacBook-Pro.hsd1.ma.comcast.net usage_softirq=0,usage_steal=0,usage_guest_nice=0,usage_system=1.9000000000005457,usage_idle=96.30000000004657,usage_nice=0,usage_iowait=0,usage_user=1.7999999999847205,usage_irq=0,usage_guest=0 1606335090000000000
swap,host=MacBook-Pro.hsd1.ma.comcast.net used_percent=75.74462890625,total=4294967296i,used=3253207040i,free=1041760256i 1606335090000000000
swap,host=MacBook-Pro.hsd1.ma.comcast.net in=0i,out=0i 1606335090000000000
swap,host=MacBook-Pro.hsd1.ma.comcast.net total=4294967296i,used=3253207040i,free=1041760256i,used_percent=75.74462890625 1606335090000000000
cpu,cpu=cpu11,host=MacBook-Pro.hsd1.ma.comcast.net usage_user=0,usage_system=0.1000000000002001,usage_iowait=0,usage_guest=0,usage_idle=99.90000000019762,usage_nice=0,usage_irq=0,usage_softirq=0,usage_steal=0,usage_guest_nice=0 1606335090000000000
swap,host=MacBook-Pro.hsd1.ma.comcast.net in=0i,out=0i 1606335090000000000
cpu,cpu=cpu12,host=MacBook-Pro.hsd1.ma.comcast.net usage_system=1.3000000000010914,usage_nice=0,usage_irq=0,usage_steal=0,usage_guest=0,usage_user=1.3999999999941792,usage_idle=97.29999999981374,usage_iowait=0,usage_softirq=0,usage_guest_nice=0 1606335090000000000
disk,device=disk1s1,fstype=apfs,host=MacBook-Pro.hsd1.ma.comcast.net,mode=ro,path=/ total=1000240963584i,free=105193611264i,used=11240935424i,used_percent=9.654295691227624,inodes_total=9767978160i,inodes_free=9767489767i,inodes_used=488393i 1606335090000000000
disk,device=disk1s2,fstype=apfs,host=MacBook-Pro.hsd1.ma.comcast.net,mode=rw,path=/System/Volumes/Data inodes_used=3848878i,total=1000240963584i,free=105193611264i,used=877618663424i,used_percent=89.29667302971013,inodes_total=9767978160i,inodes_free=9764129282i 1606335090000000000
cpu,cpu=cpu13,host=MacBook-Pro.hsd1.ma.comcast.net usage_steal=0,usage_idle=99.80000000010419,usage_nice=0,usage_softirq=0,usage_irq=0,usage_guest=0,usage_guest_nice=0,usage_user=0.09999999999792636,usage_system=0.1000000000002001,usage_iowait=0 1606335090000000000
disk,device=disk1s5,fstype=apfs,host=MacBook-Pro.hsd1.ma.comcast.net,mode=rw,path=/private/var/vm inodes_free=9767978155i,inodes_used=5i,total=1000240963584i,free=105193611264i,used=5368807424i,used_percent=4.855906272411087,inodes_total=9767978160i 1606335090000000000
cpu,cpu=cpu14,host=MacBook-Pro.hsd1.ma.comcast.net usage_user=1.5984015984033049,usage_idle=97.40259740278998,usage_nice=0,usage_iowait=0,usage_irq=0,usage_system=0.9990009990020655,usage_softirq=0,usage_steal=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu0,host=MacBook-Pro.hsd1.ma.comcast.net usage_nice=0,usage_steal=0,usage_guest=0,usage_user=10.600000000013097,usage_system=10,usage_idle=79.40000000002328,usage_guest_nice=0,usage_iowait=0,usage_irq=0,usage_softirq=0 1606335090000000000
mem,host=MacBook-Pro.hsd1.ma.comcast.net inactive=14540857344i,available=36306468864i,used=32413007872i,used_percent=47.167134284973145,available_percent=52.832865715026855,active=22328401920i,free=21765611520i,total=68719476736i,wired=5720952832i 1606335090000000000
cpu,cpu=cpu1,host=MacBook-Pro.hsd1.ma.comcast.net usage_iowait=0,usage_user=0.09999999999734427,usage_system=0.299999999998854,usage_idle=99.59999999991967,usage_nice=0,usage_irq=0,usage_softirq=0,usage_steal=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu15,host=MacBook-Pro.hsd1.ma.comcast.net usage_guest_nice=0,usage_iowait=0,usage_irq=0,usage_idle=99.89999999990715,usage_nice=0,usage_softirq=0,usage_steal=0,usage_guest=0,usage_user=0.10000000000189174,usage_system=0 1606335090000000000
cpu,cpu=cpu2,host=MacBook-Pro.hsd1.ma.comcast.net usage_idle=84.59999999981781,usage_guest=0,usage_guest_nice=0,usage_system=8.299999999975116,usage_nice=0,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_steal=0,usage_user=7.099999999970605 1606335090000000000
cpu,cpu=cpu3,host=MacBook-Pro.hsd1.ma.comcast.net usage_user=0.09999999999763531,usage_system=0.20000000000095497,usage_idle=99.70000000001164,usage_iowait=0,usage_softirq=0,usage_nice=0,usage_irq=0,usage_steal=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu4,host=MacBook-Pro.hsd1.ma.comcast.net usage_softirq=0,usage_user=5.505505505519394,usage_system=5.605605605622064,usage_idle=88.888888888986,usage_iowait=0,usage_irq=0,usage_nice=0,usage_steal=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu-total,host=MacBook-Pro.hsd1.ma.comcast.net usage_iowait=0,usage_softirq=0,usage_steal=0,usage_guest=0,usage_user=2.175679899972757,usage_system=2.2194435761167717,usage_idle=95.6048765237558,usage_nice=0,usage_guest_nice=0,usage_irq=0 1606335090000000000
cpu,cpu=cpu5,host=MacBook-Pro.hsd1.ma.comcast.net usage_nice=0,usage_irq=0,usage_softirq=0,usage_guest=0,usage_guest_nice=0,usage_user=0.0999000998979351,usage_idle=99.70029970031074,usage_steal=0,usage_system=0.1998001998004131,usage_iowait=0 1606335090000000000
cpu,cpu=cpu6,host=MacBook-Pro.hsd1.ma.comcast.net usage_nice=0,usage_irq=0,usage_guest_nice=0,usage_steal=0,usage_guest=0,usage_user=3.5964035964060654,usage_system=3.5964035964060654,usage_idle=92.80719280718787,usage_iowait=0,usage_softirq=0 1606335090000000000
cpu,cpu=cpu7,host=MacBook-Pro.hsd1.ma.comcast.net usage_guest=0,usage_guest_nice=0,usage_system=0.099999999999618,usage_idle=99.79999999981432,usage_softirq=0,usage_irq=0,usage_steal=0,usage_user=0.09999999999734427,usage_nice=0,usage_iowait=0 1606335090000000000
system,host=MacBook-Pro.hsd1.ma.comcast.net load5=2.25439453125,load15=3.0849609375,n_cpus=16i,n_users=6i,load1=2.06396484375 1606335090000000000
cpu,cpu=cpu8,host=MacBook-Pro.hsd1.ma.comcast.net usage_idle=94.20579420592753,usage_iowait=0,usage_steal=0,usage_guest=0,usage_guest_nice=0,usage_system=2.8971028970998383,usage_nice=0,usage_irq=0,usage_softirq=0,usage_user=2.897102897108924 1606335090000000000
system,host=MacBook-Pro.hsd1.ma.comcast.net uptime=453236i 1606335090000000000
cpu,cpu=cpu9,host=MacBook-Pro.hsd1.ma.comcast.net usage_irq=0,usage_softirq=0,usage_guest=0,usage_guest_nice=0,usage_nice=0,usage_system=0.10000000000047748,usage_idle=99.89999999990687,usage_iowait=0,usage_steal=0,usage_user=0 1606335090000000000
cpu,cpu=cpu10,host=MacBook-Pro.hsd1.ma.comcast.net usage_user=1.7999999999847205,usage_idle=96.30000000004657,usage_iowait=0,usage_softirq=0,usage_steal=0,usage_system=1.9000000000005457,usage_nice=0,usage_irq=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu11,host=MacBook-Pro.hsd1.ma.comcast.net usage_steal=0,usage_guest_nice=0,usage_nice=0,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_user=0,usage_system=0.1000000000002001,usage_idle=99.90000000019762,usage_guest=0 1606335090000000000
cpu,cpu=cpu12,host=MacBook-Pro.hsd1.ma.comcast.net usage_system=1.3000000000010914,usage_idle=97.29999999981374,usage_nice=0,usage_steal=0,usage_user=1.3999999999941792,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
diskio,host=MacBook-Pro.hsd1.ma.comcast.net,name=disk0 write_bytes=373449318400i,read_time=3999346i,io_time=9021701i,iops_in_progress=0i,reads=3610604i,writes=6211130i,weighted_io_time=0i,merged_reads=0i,merged_writes=0i,read_bytes=143923576832i,write_time=5022354i 1606335090000000000
cpu,cpu=cpu13,host=MacBook-Pro.hsd1.ma.comcast.net usage_nice=0,usage_iowait=0,usage_steal=0,usage_user=0.09999999999792636,usage_system=0.1000000000002001,usage_idle=99.80000000010419,usage_irq=0,usage_softirq=0,usage_guest=0,usage_guest_nice=0 1606335090000000000
cpu,cpu=cpu14,host=MacBook-Pro.hsd1.ma.comcast.net usage_softirq=0,usage_steal=0,usage_guest=0,usage_guest_nice=0,usage_system=0.9990009990020655,usage_idle=97.40259740278998,usage_iowait=0,usage_irq=0,usage_user=1.5984015984033049,usage_nice=0 1606335090000000000
system,host=MacBook-Pro.hsd1.ma.comcast.net uptime_format="5 days,  5:53" 1606335090000000000
cpu,cpu=cpu15,host=MacBook-Pro.hsd1.ma.comcast.net usage_user=0.10000000000189174,usage_system=0,usage_idle=99.89999999990715,usage_steal=0,usage_guest_nice=0,usage_nice=0,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_guest=0 1606335090000000000
cpu,cpu=cpu-total,host=MacBook-Pro.hsd1.ma.comcast.net usage_system=2.2194435761167717,usage_nice=0,usage_iowait=0,usage_guest_nice=0,usage_user=2.175679899972757,usage_idle=95.6048765237558,usage_irq=0,usage_softirq=0,usage_steal=0,usage_guest=0 1606335090000000000
diskio,host=MacBook-Pro.hsd1.ma.comcast.net,name=disk0 writes=6211130i,read_bytes=143923576832i,read_time=3999346i,write_time=5022354i,io_time=9021701i,iops_in_progress=0i,reads=3610604i,write_bytes=373449318400i,weighted_io_time=0i,merged_reads=0i,merged_writes=0i 1606335090000000000
system,host=MacBook-Pro.hsd1.ma.comcast.net load1=2.06396484375,load5=2.25439453125,load15=3.0849609375,n_cpus=16i,n_users=6i 1606335090000000000
system,host=MacBook-Pro.hsd1.ma.comcast.net uptime=453236i 1606335090000000000
system,host=MacBook-Pro.hsd1.ma.comcast.net uptime_format="5 days,  5:53" 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=en5 err_out=589i,drop_in=0i,drop_out=172i,bytes_sent=8291976i,bytes_recv=5998059i,packets_sent=41271i,packets_recv=43204i,err_in=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=en0 drop_out=1139i,bytes_sent=2727419109i,bytes_recv=3206797513i,packets_sent=6020059i,packets_recv=4436515i,err_in=0i,err_out=3837i,drop_in=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=p2p0 packets_recv=0i,err_in=0i,err_out=0i,drop_in=0i,drop_out=0i,bytes_sent=0i,bytes_recv=0i,packets_sent=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=awdl0 packets_sent=184i,packets_recv=0i,err_in=0i,err_out=0i,drop_in=0i,drop_out=0i,bytes_sent=25734i,bytes_recv=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=llw0 err_out=0i,drop_in=0i,drop_out=0i,bytes_sent=0i,bytes_recv=0i,packets_sent=0i,packets_recv=0i,err_in=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=en4 err_out=0i,drop_in=0i,drop_out=0i,bytes_sent=0i,bytes_recv=0i,packets_sent=0i,packets_recv=0i,err_in=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=en1 err_out=0i,drop_in=0i,drop_out=0i,bytes_sent=0i,bytes_recv=0i,packets_sent=0i,packets_recv=0i,err_in=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=en2 drop_out=0i,bytes_sent=0i,bytes_recv=0i,packets_sent=0i,packets_recv=0i,err_in=0i,err_out=0i,drop_in=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=en3 packets_recv=0i,err_in=0i,err_out=0i,drop_in=0i,drop_out=0i,bytes_sent=0i,bytes_recv=0i,packets_sent=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=utun0 drop_in=0i,drop_out=0i,bytes_sent=10704i,bytes_recv=0i,packets_sent=59i,packets_recv=0i,err_in=0i,err_out=0i 1606335090000000000
net,host=MacBook-Pro.hsd1.ma.comcast.net,interface=utun1 packets_recv=0i,err_in=0i,err_out=0i,drop_in=0i,drop_out=0i,bytes_sent=10704i,bytes_recv=0i,packets_sent=59i 1606335090000000000
processes,host=MacBook-Pro.hsd1.ma.comcast.net total=873i,unknown=1i,idle=0i,blocked=0i,zombies=346i,stopped=0i,running=2i,sleeping=524i 1606335090000000000
processes,host=MacBook-Pro.hsd1.ma.comcast.net blocked=0i,zombies=346i,stopped=0i,running=2i,sleeping=525i,total=873i,unknown=0i,idle=0i 1606335090000000000

```

**Comment by alamb** [comment]
I have investigated what is going on. The UI ends  asking for "tag values" and restricting the tag sets based on a `_measurement` predicate. 

It does not appear to  use the `measurement_fields` gRPC call, but instead passes a predicate like `_m == "cpu"` 

It doesn't appear to specify anywhere that it only wants field names back (does it somehow do the filtering at a higher level?) - I need to watch the gRPC requests / response from the actual storage engine more to see what is going on here. 

When I choose to see the field names in the UI, the following request goes to IOx:
<img width="1317" alt="Screen Shot 2020-11-30 at 3 56 03 PM" src="https://user-images.githubusercontent.com/490673/100663608-906f8900-3324-11eb-9da4-5b937172b32d.png">

```
[2020-11-30T20:19:41Z INFO  influxdb_iox::server::rpc::storage] tag_keys for database 26f7e5a4b7be365b_917b97a92e883afc, range: Some(TimestampRange { start: 1606763981494825000, end: 1606767581494825000 }), predicate: (TagRef:_m[0x00] == "cpu")
```
If I select both "cpu" and "disk" measurements in the UI, I see:

<img width="1297" alt="Screen Shot 2020-11-30 at 3 55 25 PM" src="https://user-images.githubusercontent.com/490673/100663576-82216d00-3324-11eb-9c63-75e32f229199.png">

```
[2020-11-30T20:20:59Z INFO  influxdb_iox::server::rpc::storage] tag_keys for database 26f7e5a4b7be365b_917b97a92e883afc, range: Some(TimestampRange { start: 1606764059181278000, end: 1606767659181278000 }), predicate: ((TagRef:_m[0x00] == "cpu") Or (TagRef:_m[0x00] == "disk"))
```

**Comment by alamb** [comment]
Here are the requests to the actual storage service (Go). I see it make both a tag_keys and a tag_values request:


```
TagKeys Request
   TagKeys source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc"
   TagKeys pred '\u0000' = \"cpu\""
   TagKeys range {1606767148115008000 1606770748115008000}"

TagKeys Response:
 0x00 (_measurement)
 cpu
 host
 0xff (_field)

TagValues
   TagValues source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc"
   TagValues pred '\u0000' = \"cpu\""
   TagValues range {1606767148172406000 1606770748172406000}"

Response:
 usage_guest
 usage_guest_nice
 usage_idle
 usage_iowait
 usage_irq
 usage_nice
 usage_softirq
 usage_steal
 usage_system
 usage_user
```

When I select two measurements in the UI, the two gRPC requests are:

```
 TagKeys
   TagKeys source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc"
   TagKeys pred '\u0000' = \"cpu\" OR '\u0000' = \"disk\""
   TagKeys range {1606767316485444000 1606770916485444000}"

Response:
 0x00 (_measurement)
 cpu
 device
 fstype
 host
 mode
 path
 0xff (_field)

TagValues
   TagValues source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc"
   TagValues pred '\u0000' = \"cpu\" OR '\u0000' = \"disk\""
   TagValues range {1606767316551837000 1606770916551837000}"

Response:
 free
 inodes_free
 inodes_total
 inodes_used
 total
 usage_guest
 usage_guest_nice
 usage_idle
 usage_iowait
 usage_irq
 usage_nice
 usage_softirq
 usage_steal
 usage_system
 usage_user
 used
 used_percent
```

**Comment by alamb** [comment]
I would summarize my finding as 

"Flux / idpe is assuming that the measurement name and field_names are encoded as tags (`_m` / 0x00, and `_f` / 0xff). This assumption is encoded into the predicates of the gRPC queries rather than in the structure of the API itself."

Given IOx has a very different model (aka it stores measurements as tables and fields as columns) this will mean we need to  recognize such predicates and handle them accordingly.

I am sure I can handle the basic case just fine, but the general case where we have arbitrary predicates in addition to tag references for `_m` and `_f` is likely to be quite slow / complicated to implement.

**Comment by alamb** [comment]
field and measurement metadata queries being encoded as predicates of the gRPC calls to tag_keys  and tag_values rather than in separate calls in the API itself.

This is using a single concept (Tags) for what are three distinct concepts in IOx:
* table name metadata (measurements)
* column name metadata(fields)
* actual tags from the data

**Comment by alamb** [comment]
More information on what is being sent back normally:

```
When I refresh the UI, here is what is sent:

```
TagKeys
   TagKeys source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc
   TagKeys pred [none]
   TagKeys range {1606832498005387000 1606836098005387000}

Response:
 0x00 (_measurement)
 cpu
 device
 fstype
 host
 interface
 mode
 name
 path
 0xff (_field)

TagValues
   TagValues source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc
   TagValues pred [none]
   TagValues range {1606832498058574000 1606836098058574000}
   TagValues tag_key \u0000

Response:
cpu
 disk
 diskio
 mem
 net
 processes
 swap
 system
```

If I click on "_field" in the UI, then I see:

```
TagValues" log_id=0Qoj7_Al000 component=read
   TagValues source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc
   TagValues pred [none]
   TagValues range {1606832791724952000 1606836391724952000}
   TagValues tag_key \ufffd

Response:
 active
 available
 available_percent
 blocked
 bytes_recv
 bytes_sent
 drop_in
 drop_out
 err_in
 err_out
 free
 idle
 in
 inactive
 inodes_free
 inodes_total
 inodes_used
 io_time
 iops_in_progress
 load1
 load15
 load5
 merged_reads
 merged_writes
 n_cpus
 n_users
 out
 packets_recv
 packets_sent
 read_bytes
 read_time
 reads
 running
 sleeping
 stopped
 total
 unknown
 uptime
 uptime_format
 usage_guest
 usage_guest_nice
 usage_idle
 usage_iowait
 usage_irq
 usage_nice
 usage_softirq
 usage_steal
 usage_system
 usage_user
 used
 used_percent
 weighted_io_time
 wired
 write_bytes
 write_time
 writes
 zombies
```

Which basically means "all field across all measurements"
```

**Comment by alamb** [comment]
So after some more digging, it is stranger and stranger. 

When I click on the `_measurement` = CPU, as described above, the UI sends this query:

```
from(bucket: "devbucket")
|> range(start: -1h, stop: now())
|> filter(fn: (r) => true)
|> keep(columns: ["_field"])
|> group()
|> distinct(column: "_field")
|> limit(n: 200)
|> sort()"
```

When the actual storaged engine is hooked up,  I see the two gRPC requests: 

```
TagKeys Request
   TagKeys source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc"
   TagKeys pred '\u0000' = \"cpu\""
   TagKeys range {1606767148115008000 1606770748115008000}"

TagKeys Response:
 0x00 (_measurement)
 cpu
 host
 0xff (_field)

TagValues
   TagValues source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc"
   TagValues pred '\u0000' = \"cpu\""
   TagValues range {1606767148172406000 1606770748172406000}"
   TagValues tag_key \ufffd"

Response:
 usage_guest
 usage_guest_nice
 usage_idle
 usage_iowait
 usage_irq
 usage_nice
 usage_softirq
 usage_steal
 usage_system
 usage_user
```

However, when I have IOx hooked up, the flux query is still sent but an empty response is returned (and IOx only sees a single request), not a tag_values request too

```
[2020-12-01T16:10:58Z INFO  influxdb_iox::server::rpc::storage] AAL tag keys 
[2020-12-01T16:10:58Z INFO  influxdb_iox::server::rpc::storage] tag_keys for database 26f7e5a4b7be365b_917b97a92e883afc, range: Some(TimestampRange { start: 1606835458953316000, end: 1606839058953316000 }), predicate: (TagRef:_m[0x00] == "cpu") 
[2020-12-01T16:10:58Z INFO  influxdb_iox::server::rpc::storage] Returning tag keys 
[2020-12-01T16:10:58Z INFO  influxdb_iox::server::rpc::storage]    
[2020-12-01T16:10:58Z INFO  influxdb_iox::server::rpc::storage]   cpu 
[2020-12-01T16:10:58Z INFO  influxdb_iox::server::rpc::storage]   host 
[2020-12-01T16:10:58Z INFO  influxdb_iox::server::rpc::storage]   � 
```
🤔

**Comment by alamb** [comment]
My next plan will be to go spelunking in the flux code to see if I can figure out why in some cases it is sending a `tag_values` request and in some cases it does not

**Comment by alamb** [comment]
Digging a little more, when the "cpu" measurement is clicked in the UI, it sends two queries to the backend:

```
query: "from(bucket: "devbucket")
  |> range(start: -1h, stop: now())
  |> filter(fn: (r) => (r["_measurement"] == "cpu"))
  |> keys()
  |> keep(columns: ["_value"])
  |> distinct()
  |> filter(fn: (r) => r._value != "_measurement")
  |> filter(fn: (r) => r._value != "_time" and r._value != "_start" and r._value !=  "_stop" and r._value != "_value")
  |> sort()
  |> limit(n: 200)"
````

With results: 

```
#group,false,false,false
#datatype,string,long,string
#default,_result,,
,result,table,_value
,,0,_field
,,0,cpu
,,0,host
```

And the second query sent by the UI is the same as above
```
query: "from(bucket: "devbucket")
  |> range(start: -1h, stop: now())
  |> filter(fn: (r) => (r["_measurement"] == "cpu"))
  |> keep(columns: ["_field"])
  |> group()
  |> distinct(column: "_field")
  |> limit(n: 200)
  |> sort()"
```

When storage / TSM2 is the backend, the results are:
```
#group,false,false,false
#datatype,string,long,string
#default,_result,,
,result,table,_value
,,0,usage_guest
,,0,usage_guest_nice
,,0,usage_idle
,,0,usage_iowait
,,0,usage_irq
,,0,usage_nice
,,0,usage_softirq
,,0,usage_steal
,,0,usage_system
,,0,usage_user
```

When IOx is the backend, the results are:
```
#group,false,false,false
#datatype,string,long,string
#default,_result,0,
,result,table,_value

```

**Comment by alamb** [comment]
🤔  So I fired up wireshark and watched what was going on port 8082 (the port that gRPC is running on). 

I am no wireshark expert, but by my limited reading it looks to me like the IOx server is actually getting a second request, but just not passing that request along (nor making any error logs...)

I am going to investigate that angle a little more

![Screen Shot 2020-12-01 at 4 38 10 PM](https://user-images.githubusercontent.com/490673/100800061-055dc400-33f4-11eb-8c2c-f19601d3d920.png)

**Comment by alamb** [comment]
Ok, I figured it out!

The root cause of the issue is that the `tag_value` request is (techncially) invalid as it includes a `0xff` value in a protobuf `string` field. This is the same basic problem as on https://github.com/influxdata/influxdb_iox/issues/346 but in a different message.

Specifically, here is the `tag_values` request that is made to IOx:

```
TagValues" log_id=0Qoj7_Al000 component=read
   TagValues source partition_id: 0 org_id: 26f7e5a4b7be365b bucket_id: 917b97a92e883afc
   TagValues pred [none]
   TagValues range {1606832791724952000 1606836391724952000}
   TagValues tag_key \ufffd
```

The `tag_key` field is of `string` type, but in this case it has the value `0xff` (the magic "_field" special tag value). However, since 0xff is not a valid UTF-8 string and Rust barfs on it.

That is a relatively easy fix and I'll get it coded up shortly.


Most troubling is that the errors got swallowed and I spent several hours trying to figure out where the problem was (IOx has no record of the request in the log). I putzed for a while trying to get a hook high enough up the tonic/hyper/prost stack to record that we at least got a request and was unsuccessful.

I have filed https://github.com/influxdata/influxdb_iox/issues/514 to track this problem
